Add a test with a cpp executable including all mbed TLS headers

AndrzejKurek opened the pull request:
Related to #1426 - in order to avoid such mistakes in the future, a test with all the mbed TLS headers included during a c++ executable building is added.

Internal ref: IOTSSL-2172

AndrzejKurek commented:
And we've got a winner:
> /home/travis/build/ARMmbed/mbedtls/programs/test/header_test.cpp:117:2: error: expected '}'
> }
>  ^
> /home/travis/build/ARMmbed/mbedtls/include/mbedtls/rsa_internal.h:63:12: note: to match this '{'
> extern "C" {

gilles-peskine-arm commented on programs/Makefile:
```diff
@@ -67,6 +69,7 @@ APPS =	aes/aescrypt2$(EXEXT)		aes/crypt_and_hash$(EXEXT)	\
 	random/gen_random_ctr_drbg$(EXEXT)				\
 	test/ssl_cert_test$(EXEXT)	test/benchmark$(EXEXT)		\
 	test/selftest$(EXEXT)		test/udp_proxy$(EXEXT)		\
+	test/header_test$(EXEXT)								\
```
I don't think we should build this program by default. People who only have a C compiler and no C++ compiler, or who've set `CC` but not `CXX`, should be able to run `make`.

This applies to other build systems as well.

AndrzejKurek replied:
Fixed.

gilles-peskine-arm commented on programs/test/header_test.cpp:
```diff
@@ -0,0 +1,117 @@
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_printf     printf
```
Why bother with these definitions here?

AndrzejKurek replied:
Just wanted to keep the same include format as in the other mbed TLS files, but I guess it's not necessary here for now. Removed it.

AndrzejKurek commented:
This feature can be enabled by either:
- Defining TEST_CPP using makefiles - make TEST_CPP=1
- Defining TEST_CPP using cmake - (running cmake from a separate folder in mbed TLS directory, just to be tidy) cmake -D CMAKE_BUILD_TYPE:String="Check" -D TEST_CPP="yes" ../

k-stachowiak commented on programs/test/header_test.cpp:
```diff
@@ -0,0 +1,117 @@
+
+int main( int argc, char *argv[] )
+{
+    (void) argc;
```
C standard allows main to take no arguments, which seems to be exactly what is needed here.

AndrzejKurek replied:
Good tip, done.

mazimkhan commented on programs/test/header_test.cpp:
```diff
@@ -0,0 +1,109 @@
+int main( int argc, char *argv[] )
+{
+    (void) argc;
+    (void) argv;
```
Missing ```return``` in a non void function.

gilles-peskine-arm replied:
In C++, that's allowed in `main`, by special dispensation. `main` has an implicit `return 0`.

mazimkhan commented on programs/test/header_test.cpp:
```diff
@@ -0,0 +1,109 @@
+#include "mbedtls/timing.h"
+#include "mbedtls/x509.h"
+
+#if defined(MBEDTLS_PLATFORM_C)
```
Checking ```MEBDTLS_x_C``` should be done for all modules. Or if we always build this test with full config then why check here. For any reason if implementation for a header is missing, stubs can be provided in this test completeness sake.

gilles-peskine-arm replied:
You can include any header file in any configuration, it's up to each header file to exclude things that shouldn't be included in a given configuration. `MBEDTLS_PLATFORM_C` is special in that regard.

AndrzejKurek replied:
This file was intended for our CI and mainly for the cmake build, but I left a possibility for people to test their configuration with the most common exclusions - custom config.h and custom platform macros. Do you think that I should change this code, if so - how?

AndrzejKurek replied:
And - are you concerned about Including all headers despite having a smaller configuration? There are no guards in the headers anyway, .c files have ifdef guards.

mazimkhan replied:
I known that the guards are in C files. My question was why guard ```platform.h``` unlike others.

mazimkhan commented on programs/Makefile:
```diff
@@ -241,6 +247,10 @@ test/benchmark$(EXEXT): test/benchmark.c $(DEP)
 	echo "  CC    test/benchmark.c"
 	$(CC) $(LOCAL_CFLAGS) $(CFLAGS) test/benchmark.c   $(LOCAL_LDFLAGS) $(LDFLAGS) -o $@
 
+test/header_test$(EXEXT): test/header_test.cpp $(DEP)
```
To make this test self sufficient perhaps execute config.pl with full config.

gilles-peskine-arm replied:
Good idea, but that's up to test scripts such as `all.sh`, not the job of the build scripts.

AndrzejKurek replied:
What do you mean by self sufficient?

mazimkhan replied:
My idea was to have a make target that indicates whether the headers are compatible with C++. But general consensus here is do the config outside Makefile. Hence ignore this.

mazimkhan commented on programs/Makefile:
```diff
@@ -76,6 +78,10 @@ ifdef PTHREAD
 endif
 
+ifdef TEST_CPP
+APPS += test/header_test$(EXEXT)
```
Since this test tests C++ compatibility of the headers with full config. It is different from other tests/programs that all build with one valid config. Please check if a full config is valid for other programs.
If not I would rather make it mutually exclusive with all other programs.

AndrzejKurek replied:
This tests headers with any config - there are usually no ifdef guards in headers, hence nothing really should change.

sbutcher-arm replied:
In `all.sh` we should enable the full config and build this. The actual `Makefile` should just take the configs it's given.

sbutcher-arm commented on programs/.gitignore:
```diff
@@ -45,6 +45,7 @@ ssl/mini_client
 test/benchmark
 test/ecp-bench
 test/selftest
+test/header_test
```
I don't think `header_test` really conveys what we want to do here. Can we call it something like `cpp_dummy_build`? Or at least something that says it's a C++ test.

AndrzejKurek replied:
Done - renamed to cpp_dummy_build.

sbutcher-arm commented:
Note fix relates to existing bug #477 (IOTSSL-761), which should be closed when we merge this.

sbutcher-arm commented on programs/test/header_test.cpp:
```diff
@@ -0,0 +1,109 @@
+/*
+ *  A C++ program that includes all of the mbed TLS header files, in order to
+ *  test if no errors are raised in the process.
+ *
```
Rather than say we want to check we build without errors...
~~~
/*
 *  A C++ program that includes all of the mbed TLS header files, in order to
 *  test if no errors are raised in the process.
~~~
I'd like to state the purpose more clearly, so I'd propose:
~~~
/*
 * This program is a dummy C++ program to ensure Mbed TLS library header files
 * can be included and built with a C++ compiler.
~~~

AndrzejKurek replied:
Done, updated as suggested.

sbutcher-arm commented on programs/test/header_test.cpp:
```diff
@@ -0,0 +1,109 @@
+#include "mbedtls/memory_buffer_alloc.h"
+#endif
+
+int main( int argc, char *argv[] )
```
Ok - this is probably a naive question, but is this adequate? Shouldn't we want to check it links with a token function? 

Mbed TLS works with C++ by compiling in C (not C++) as a C library, which the C++ linker can then resolve without the usual name mangling. As a build test, we're only confirming the preprocessor and compiler work - not the linker.

Can I propose we should call `mbedlts_platform_setup()` and then `mbedtls_printf("CPP Build test\n")` as a simple sanity test? (And the `mbedtls_platform_teardown()` for completeness).

gilles-peskine-arm replied:
If we want to check this properly, we need to link with _every_ function in the library. This would detect a missing `extern "C"` somewhere.

This means the `.cpp` source file would have to be automatically generated, which requires a rather different approach.

Automatically generating the source file would have the advantage that the list of headers would also be automatically generated. With a static file, we have to remember to add new headers. Fortunately this doesn't happen often. Still, we should add a check script (or code in an existing script) to verify that all the headers are included.

sbutcher-arm replied:
I think trying to link every API function and/or a script to confirm every module is included is too much work for the moment, and a distraction from other issues. Feel free to raise those as issues if you want for the future.

However, I think we can link to a couple of files as I suggest above at little extra cost, and at least that would confirm the linker is working for something.

AndrzejKurek replied:
Added a basic set of calls as suggested.

AndrzejKurek commented:
Sorry for marking as ready for merge - I thought that 2 approvals are enough for that.

AndrzejKurek commented:
Do you think that there should be a Visual Studio test? The C++ header check is mostly a sanity check, with a possibility for users to perform a check using their own set of headers (included via config.h). Current Visual Studio builds (as we're using it) create a C project, building our library as a C library, so creating a C++ Visual Studio project would be a new feature, with different guards to be added (for people that only want C Visual Studio builds).

mpg commented:
Re-adding "needs review" + "needs work" as there is one "change requested" review.

sbutcher-arm commented:
@AndrzejKurek - Looks like there's a small amount of rework and then we're good. Could you please look at this please? If you don't have access to Visual Studio, please let me know and we'll do something else.

AndrzejKurek commented:
@sbutcher-arm Please see my comment about the C++ test - this would be a new feature.
@gilles-peskine-arm - Please reply to my comment.

sbutcher-arm commented:
@AndrzejKurek - I'm satisfied by your point, and am happy to leave the PR as is. I need to know @gilles-peskine-arm's feedback before I approve for merge however.

sbutcher-arm commented:
This PR failed the Jenkins CI:
```
******************************************************************
* test: check-files.py 
* Fri Jun 22 14:22:31 UTC 2018
******************************************************************
Tabs present:
./CMakeLists.txt: 3, 5
./programs/test/CMakeLists.txt: 20, 21
```

@AndrzejKurek - Can you please fix this, and while you're reworking it, can you please add a test for this into to `all.sh`, either in this PR or as a new PR? Thanks.

AndrzejKurek commented:
I've fixed the tab issues and updated the PR with two additional tests in all.sh - one using make, and one using cmake to run the CXX dummy compilation. These are enabled by default and can be disabled with --no-cxx flag. @gilles-peskine-arm, @mazimkhan, @sbutcher-arm, @k-stachowiak - please re-review.

sbutcher-arm commented on tests/scripts/all.sh:
```diff
@@ -127,6 +129,7 @@ General options:
      --armcc            Run ARM Compiler builds (on by default).
      --no-armcc         Skip ARM Compiler builds.
      --no-yotta         Skip yotta module build.
+     --no-cxx           Skip CXX Compiler build.
```
Why might someone want to skip the C++ build? I don't see a reason for having an option.

gilles-peskine-arm replied:
`make` shouldn't run C++ by default because there are C compilers without an associated C++ compilers. But `all.sh` assumes that you have GCC, which has a C++ compiler. So I agree, this doesn't need to be optional in `all.sh`.

AndrzejKurek replied:
I'd rather assume that there are targets that don't have a C++ compiler in the toolchain. It may happen for some embedded targets that have their toolchains generated (and stripped off any unnecessary tools). I just wanted to make it possible for these people to turn the support off with ease, hoping that giving more possibilities is better. This test will run by default by the way.

mpg replied:
As Gilles said, this would be a relevant argument if we were talking about `make`, but here we're talking about `all.sh`, which doesn't build for arbitrary targets: it builds for a very specific set of targets with a very specific set of toolchains (namely gcc, clang, arm-gcc, armcc5 and armclang). And we specify the exact toolchain on each `make` invocation. This isn't a general-purpose script, it already requires a very specific environment (in addition to the specific toolchains, specific version of OpenSSL and GnuTLS, and so on), and is not meant to be adapted to everyone's needs: it's just for us and the CI.

We should take advantage of the fact that this script is not generic and avoid any unnecessary complexity.

sbutcher-arm commented:
I hate to be the one to point it out, but you have conflicts, and this PR likely needs rebasing.

Also - I wasn't sure why we have the extra option in `all.sh`. I can't see why we'd need it.

k-stachowiak commented on tests/scripts/all.sh:
```diff
@@ -557,6 +563,17 @@ cleanup
+if [ $TEST_CXX -ne 0 ]; then
+    msg "build: Unix make, gcc and g++ test" # ~ 30s
+    cleanup
+    make TEST_CPP=1
```
Is there a reason for this being `CPP` rather than `CXX` like elsewhere? It's far from being a blocker, but it is possible to make it consistent, I'd take the opportunity.

gilles-peskine-arm replied:
Agreed, CPP is the C preprocessor.

AndrzejKurek replied:
Sure, I can change it, I just used it as an abbreviation of C++ (as in cppreference, not cxxreference).

AndrzejKurek replied:
Update - Removed the flag alltogether.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -557,6 +563,17 @@ cleanup
+
+    msg "build: cmake, gcc and g++ test" # ~ 30s
+    cleanup
+    CC=gcc cmake -D TEST_CPP=YES .
```
Why test with both make and cmake? We run some tests with each to validate our build system. But here we're just building a program which has no purpose other than validating that it can be built. Any build system would do.

AndrzejKurek replied:
I opted for a larger coverage of build systems, just in case someone forces the CXX compiler for cmake to something not working.

AndrzejKurek replied:
Update - Removed the cmake test.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -557,6 +563,17 @@ cleanup
+
+    msg "build: cmake, gcc and g++ test" # ~ 30s
+    cleanup
+    CC=gcc cmake -D TEST_CPP=YES .
```
Does `CC=gcc` influence the _C++_ compiler? Seems strange to me, I'd expect the C++ compiler to be `${CXX}`.

AndrzejKurek replied:
It does not, It's just the usual syntax used for other tests in all.sh, and I thought it's a prerequisite to compile the usual sources (which are getting compiled during this test too).

mpg commented:
Just a general note on this PR: when it is merge, there should be an announcement on the relevant channels so that all developers know that from now on, every time you create a new module, you also new to add it's header to the C++ testing file.

AndrzejKurek commented:
I have rebased this PR, incorporated all of the comments, thus making the cpp test mandatory in all.sh (with no flag for removal), removed the cmake test in all.sh.
I have also updated the header list according to head state, and I've reordered them in an alphabetical order, as it is easier to spot any inconsistencies this way. 
@k-stachowiak , @sbutcher-arm , @mazimkhan , @gilles-peskine-arm - please re-review.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -580,6 +581,10 @@ msg "test/build: key-exchanges (gcc)" # ~ 1 min
 
+msg "build: Unix make, gcc and g++ test" # ~ 30s
+cleanup
+make TEST_CPP=1
```
Should we merge this with an existing build?

Advantage: faster.

Downside: slightly less good reports if it fails since we'd have to look closer at the logs to see if it's the fault of the C++. But this is `all.sh` anyway, we don't.

I'm not suggesting to add `TEST_CPP=1` to an existing command line, that would make it harder to backport this  script. Rather, add this after an existing `make`, before a subsequent `cleanup`.

AndrzejKurek replied:
So you'd like to have the following code below:
```
 msg "build: Unix make, -Os (gcc)" # ~ 30s
 cleanup
 make CC=gcc CFLAGS='-Werror -Wall -Wextra -Os'
+make TEST_CPP=1
```
?

gilles-peskine-arm replied:
Yes, that's fine with me.

AndrzejKurek replied:
Updated, with additional logs to make things clear.

gilles-peskine-arm commented on programs/test/cpp_dummy_build.cpp:
```diff
@@ -0,0 +1,117 @@
+#include "mbedtls/md2.h"
+#include "mbedtls/md4.h"
+#include "mbedtls/md5.h"
+#include "mbedtls/md.h"
```
Thanks for sorting this list, but I'd prefer if it was sorted in binary order, rather than in some locale where punctuation is not sorted or sorted in a second pass and in a non-ASCII order (`md.h` should come before `md2.h`, `ssl.h` should come before `ssl_anything.h`).

gilles-peskine-arm replied:
Please use the `C` locale for universally reproducible results.

AndrzejKurek replied:
Done.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -580,6 +581,10 @@ msg "test/build: key-exchanges (gcc)" # ~ 1 min
+msg "build: Unix make, gcc and g++ test" # ~ 30s
+cleanup
+make TEST_CPP=1
+
```
I would like to add a check that `cpp_dummy_build.cpp` lists all the headers except for whitelisted ones.
```
check_headers_in_cpp () {
    ls include/mbedtls >headers.txt
    <programs/test/cpp_dummy_build.cpp sed -n 's/"$//; s!^#include "mbedtls/!!p' |
    sort |
    diff headers.txt -
    rm headers.txt
}
record_status check_headers_in_cpp
```

AndrzejKurek replied:
Will there be any whitelisted ones? Currently, this fails only because check-config.h is not included there (there's no need to test it since it does not have any extern directives), but there's no problem in adding it there.

gilles-peskine-arm replied:
`check_config.h` is the only whitelisted one.

AndrzejKurek replied:
Added a check, included the check-config.h in the dummy build file.

AndrzejKurek commented:
@gilles-peskine-arm, @k-stachowiak, @mazimkhan - Could I kindly ask for your re-review?

gilles-peskine-arm commented on programs/Makefile:
```diff
@@ -77,6 +79,10 @@ ifdef PTHREAD
 endif
 
+ifdef TEST_CPP
+APPS += test/cpp_dummy_build$(EXEXT)
```
If you run `make TEST_CPP=1` and later `make clean`, then `test/cpp_dummy_build` doesn't get cleaned. There's a similar preexisting defect with `ssl/ssl_pthread_server`. The defect with `test/cpp_dummy_build` is worse because we include it in our test scripts.

I won't block the PR for that, but please, in order of preference, either:

* arrange for `make clean` to remove `test/cpp_dummy_build` (and `ssl/ssl_pthread_server`) regardless of the value of `TEST_CPP`;
* or file an issue.

sbutcher-arm replied:
@AndrzejKurek is out for a couple of weeks, so I've raised this as issue #1862, so we can at least merge what we have, and fix this later, as it's a minor issue,

sbutcher-arm commented:
CircleCI is failing, but that can be disregarded.
